### DISCUSSION TITLE
Export IfcGridAxis data using IfcConvert

### DISCUSSION THREAD

**Discussion by rosterplayer** [IfcOpenShell/IfcOpenShell#5361] [Q&A] [answered]
Good day!
I want to export coordinates of IfcGridAxis to json file using IfcConvert. Try to add some functions in IfcConvert.cpp. Here is some code:
        `template <typename Schema>
	json get_axis(boost::shared_ptr<typename Schema::IfcGridAxis::list> axis, IfcParse::IfcFile& f)
	{
		json j_axes = json::array();
		for (auto& rel : *axis) {

			auto castedAxis = dynamic_cast<typename Schema::IfcGridAxis*>(rel);
			auto axisCurve = castedAxis->AxisCurve();

			boost::optional<std::string> axisName = castedAxis->AxisTag();

			Handle(Geom_Curve) curve;

                        // Any way to make this scheme agnostic?
                        //IfcGeom::KernelIfc4 kernel;
			//IfcGeom::KernelIfc2x3 kernel;

			gp_Pnt startPoint;
			gp_Pnt endPoint;

			TopoDS_Wire wire;

			if (kernel.convert_curve(axisCurve, curve))
			{
				auto start = curve->FirstParameter();
				auto end = curve->LastParameter();
				startPoint = curve->Value(start);
				endPoint = curve->Value(end);
			}
			else if (kernel.convert_wire(axisCurve, wire))
			{
	#if OCC_VERSION_HEX < 0x70600
				BRepAdaptor_CompCurve cc(wire, true);
				Handle(Adaptor3d_HCurve) hcc = Handle(Adaptor3d_HCurve)(new BRepAdaptor_HCompCurve(cc));
	#else
				auto hcc = new BRepAdaptor_CompCurve(w, true);
	#endif

				auto start = hcc->FirstParameter();
				auto end = hcc->LastParameter();
				startPoint = hcc->Value(start);
				endPoint = hcc->Value(end);
			}

			...`

I'm new to c++ and my problem how to make this function schema independent to use convert_curve and convert_wire methods?

**Comment by rosterplayer** [comment, marked as answer]
Solved the problem by adding convert_curve() and convert_wire() to Kernel and making them virtual

**Comment by aothms** [comment]
Ok, interesting. Glad you managed to resolve it. Typically the grid axes curves are also part of the representation of the grid directly so with `--model --plan` added to convert you get them in the geometric output, but if you need a more structured output in JSON I guess this is the only way.
